rust/kernel/str.rs

```diff
@@ -0,0 +1,225 @@
+        }
+        if bytes[..bytes.len()].contains(&0) {
+            return Err(CStrConvertError::InteriorNul);
+        }
+        // SAFETY: We just checked that all properties hold.
+        Ok(unsafe { Self::from_bytes_with_nul_unchecked(bytes) })
+    }
+
+    /// Creates a [`CStr`] from a `[u8]` without performing any additional
+    /// checks.
+    ///
+    /// # Safety
+    ///
+    /// `bytes` *must* end with a `NUL` byte, and should only have a single
+    /// `NUL` byte (or the string will be truncated).
+    #[inline]
+    pub const unsafe fn from_bytes_with_nul_unchecked(bytes: &[u8]) -> &CStr {
+        // Note: This can be done using pointer deref (which requires
+        //       const_raw_ptr_deref to be const) or transmute (which requires
+        //       const_transmute to be const) or ptr::from_raw_parts (which
+        //       requires ptr_metadata).
+        //       While none of them are current stable, it is very likely that
+        //       one of them will eventually be.
+        &*(bytes as *const [u8] as *const Self)
+    }
+
+    /// Returns a C pointer to the string.
+    #[inline]
+    pub const fn as_char_ptr(&self) -> *const c_types::c_char {
+        self.0.as_ptr() as _
+    }
+
+    /// Convert the string to a byte slice without the trailing 0 byte.
+    #[inline]
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0[..self.0.len() - 1]
+    }
+
+    /// Convert the string to a byte slice containing the trailing 0 byte.
+    #[inline]
+    pub const fn as_bytes_with_nul(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl AsRef<BStr> for CStr {
+    #[inline]
+    fn as_ref(&self) -> &BStr {
+        self.as_bytes()
+    }
+}
+
+impl Deref for CStr {
+    type Target = BStr;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_bytes()
+    }
+}
+
+impl Index<ops::RangeFrom<usize>> for CStr {
+    type Output = CStr;
+
+    #[inline]
+    fn index(&self, index: ops::RangeFrom<usize>) -> &Self::Output {
+        assert!(
+            index.start <= self.len(),
+            "range start index {} out of range for slice of length {}",
+            index.start,
+            self.len()
+        );
+        // SAFETY: We just checked the length.
+        unsafe { Self::from_bytes_with_nul_unchecked(&self.0[index.start..]) }
+    }
+}
+
+impl Index<ops::RangeFull> for CStr {
+    type Output = CStr;
+
+    #[inline]
+    fn index(&self, _index: ops::RangeFull) -> &Self::Output {
+        self
+    }
+}
+
+#[doc(hidden)]
+pub trait CStrIndex {}
+
+impl CStrIndex for usize {}
+impl CStrIndex for ops::Range<usize> {}
+impl CStrIndex for ops::RangeInclusive<usize> {}
+impl CStrIndex for ops::RangeToInclusive<usize> {}
+
+impl<Idx> Index<Idx> for CStr
+where
+    Idx: CStrIndex,
+    BStr: Index<Idx>,
+{
+    type Output = <BStr as Index<Idx>>::Output;
+
+    #[inline]
+    fn index(&self, index: Idx) -> &Self::Output {
+        &self.0[index]
```

Comment:
This doesn't check that the resulting `CStr` still ends with a null terminator. For `RangeInclusive` and `RangeToInclusive` this will likely not be the case most of the time. I think only `RangeFrom` and `usize` are fine to index a `CStr` with.

Reply:
The output of this operation is a `BStr` instead of `CStr`, so the output don't need to be NUL-terminated. But the code is indeed wrong and it should actually be `&self.as_bytes()[index]`.